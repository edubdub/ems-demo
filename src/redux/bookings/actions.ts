import { createAction } from 'redux-actions'
import { List } from 'immutable'
import { Booking } from '../../models/booking'

export const actions = {
  LOAD_BOOKINGS: 'LOAD_BOOKINGS',
  SET_BOOKINGS: 'SET_BOOKINGS',
  SET_SEARCH_TERMS: 'SET_SEARCH_TERMS',
  SET_USE_FUZZY_SEARCH: 'SET_USE_FUZZY_SEARCH',
  ADD_NEW_EVENT: 'ADD_NEW_EVENT'
}

export const actionCreators = {
  loadBookings: createAction(actions.LOAD_BOOKINGS, () => null),
  setBookings: createAction(actions.SET_BOOKINGS, (payload: Booking[]|List<Booking>) => List(payload)),
  setSearchTerms: createAction(actions.SET_SEARCH_TERMS, (payload: string) => payload),
  setUseFuzzySearch: createAction(actions.SET_USE_FUZZY_SEARCH, (payload: boolean) => payload)
}
