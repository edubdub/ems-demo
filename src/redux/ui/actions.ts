import { createAction } from 'redux-actions'
export const actions = {
  SET_NAV_BAR_HEIGHT: 'SET_NAV_BAR_HEIGHT',
  SET_BOOKING_DATE_WINDOW_POSITION: 'SET_BOOKING_DATE_WINDOW_POSITION',
  USER_SET_NAV_DATE: 'USER_SET_NAV_DATE',
  SET_NAV_DATE: 'SET_NAV_DATE',
  SET_SHOULD_NAVIGATE: 'SET_SHOULD_NAVIGATE'
}

export const actionCreators = {
  setNavBarHeight: createAction(actions.SET_NAV_BAR_HEIGHT, (height: number) => height),
  setBookingDateWindowPosition: createAction(
    actions.SET_BOOKING_DATE_WINDOW_POSITION,
    (bookingDate: string, position: number) => ({
      bookingDate, position
    })),
  userSetNavDate: createAction<Date>(actions.USER_SET_NAV_DATE),
  setNavDate: createAction<Date>(actions.SET_NAV_DATE)
}
