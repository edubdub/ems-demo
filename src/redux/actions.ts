import {actionCreators as bookings } from './bookings/actions'
export default {
  bookings
}
