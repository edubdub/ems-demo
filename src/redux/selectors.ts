import bookings from './bookings/selectors'
export default {
  bookings
}
