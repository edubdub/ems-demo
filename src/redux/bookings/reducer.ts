import { handleActions, Action } from 'redux-actions'
import { actions } from './actions'
import { Record, List } from 'immutable'
import { Booking } from '../../models/booking'
const initialState = {
  loading: false,
  bookings: List<{}>(),
  searchTerms: '',
  useFuzzySearch: true,
  addEventOpen: false
}
export class BookingState extends Record(initialState) {
  loading: boolean
  bookings: List<Booking>
  searchTerms: string
  useFuzzySearch: boolean
  addEventOpen: false
}
const defaultState = new BookingState(initialState)
const reducer = handleActions({
  [actions.LOAD_BOOKINGS](state: BookingState) {
    return state.set('bookings', List()).set('loading', true)
  },
  [actions.SET_BOOKINGS](state: BookingState, action: Action<Booking[]>) {
    return state.set('bookings', action.payload).set('loading', false)
  },
  [actions.SET_SEARCH_TERMS](state: BookingState, action: Action<string>) {
    return state.set('searchTerms', action.payload)
  },
  [actions.SET_USE_FUZZY_SEARCH](state: BookingState, action: Action<boolean>) {
    return state.set('useFuzzySearch', action.payload)
  }
} as any, defaultState)

export default reducer
