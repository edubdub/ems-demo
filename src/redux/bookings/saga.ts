import { takeLatest, call, put } from 'redux-saga/effects'
import { actions, actionCreators } from './actions'
import * as bookingService from '../../services/bookings'
export function * bookingSaga() {
  yield takeLatest(actions.LOAD_BOOKINGS, loadBookings)

}

export function * loadBookings () {
  const bookings = yield call(bookingService.requestBookings)
  yield put(actionCreators.setBookings(bookings))
}
