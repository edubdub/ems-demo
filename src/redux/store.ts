import { createStore, applyMiddleware, combineReducers, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import bookingReducer, { BookingState } from './bookings/reducer'
import uiReducer, { UIState } from './ui/reducer'
import { bookingSaga } from './bookings/saga'
export interface IState {
  bookings: BookingState,
  ui: UIState
}

const reducer = combineReducers({
  bookings: bookingReducer,
  ui: uiReducer
})

const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
const sagaMiddleware = createSagaMiddleware()
export const store = createStore(
  reducer,
  composeEnhancers(
    applyMiddleware(
      sagaMiddleware
    )
  )

)
sagaMiddleware.run(bookingSaga)
