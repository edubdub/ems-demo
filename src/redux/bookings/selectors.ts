import { createSelector } from 'reselect'
import { IState } from '../store'
import { Booking } from '../../models/booking'
import * as Fuse from 'fuse.js'
import beginningOfDay from '../../utils/date/beginningOfDay'
const fuseOptions = {
  shouldSort: false,
  threshold: 0.6,
  location: 0,
  distance: 100,
  maxPatternLength: 32,
  minMatchCharLength: 1,
  keys: [
    'eventName',
    'roomName'
  ]
}

const loading = (state: IState) => state.bookings.loading
const useFuzzySearch = (state: IState) => state.bookings.useFuzzySearch
const allBookings = (state: IState) => state.bookings.bookings
const searchableBookings = createSelector(allBookings, useFuzzySearch, (bookings, fs) => {
  if (fs) {
    return new Fuse(bookings.toArray(), fuseOptions)
  } else {
    return {
      search: (term: string) => bookings
        .toArray()
          .filter((b) => b &&
            b.eventName.toLowerCase().match(term.toLowerCase()) ||
            b.roomName.toLowerCase().match(term.toLowerCase()))
    }
  }
})
const currentSearchTerm = (state: IState) => state.bookings.searchTerms
const selectBookings = createSelector(
  searchableBookings,
  currentSearchTerm,
  allBookings,
  (searchableBookings, searchTerm, bookings) => searchTerm ? searchableBookings.search(searchTerm) : bookings)
const bookingsByDate = createSelector(selectBookings, (bookings: Array<Booking>) => {
  return bookings.reduce((map, b) => {
    const bookingDate = beginningOfDay(b.start).toISOString()
    map[bookingDate] = map[bookingDate] || []
    map[bookingDate].push(b)
    return map
  }, {})
})
export default {
  allBookings,
  selectBookings,
  loading,
  bookingsByDate,
  useFuzzySearch,
  currentSearchTerm
}
