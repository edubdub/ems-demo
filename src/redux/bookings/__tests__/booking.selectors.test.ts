import { getStore } from '../../store'
import selectors from '../selectors'
import { data } from '../../../services/bookings'
import { actionCreators } from '../actions'
describe('selectors', () => {
  describe('with default states', () => {
    let store = getStore()
    beforeEach(() => {
      store = getStore()
    })
    Object.keys(selectors).forEach(key => {
      it('should be the same with ' + key, () => {
        expect(selectors[key](store.getState())).toMatchSnapshot()
      })
    })
  })
  describe('with bookings', () => {
    let store = getStore()
    beforeEach(() => {
      store = getStore()
      store.dispatch(actionCreators.setBookings(data))
    })
    Object.keys(selectors).forEach(key => {
      it('should be the same with ' + key, () => {
        expect(selectors[key](store.getState())).toMatchSnapshot()
      })
    })
  })
  // TODO: this could use more unit tests, but i'm out of time, particularly the selectors for nearest dates
})
