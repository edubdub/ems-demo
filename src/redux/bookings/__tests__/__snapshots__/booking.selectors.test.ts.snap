// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors with bookings should be the same with allBookings 1`] = `
Immutable.List [
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 01",
    "id": 0,
    "roomName": "Demo Room 01",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 02",
    "id": 1,
    "roomName": "Demo Room 02",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 03",
    "id": 2,
    "roomName": "Demo Room 03",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 04",
    "id": 3,
    "roomName": "Demo Room 04",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 05",
    "id": 4,
    "roomName": "Demo Room 05",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Thanksgiving",
    "id": 5,
    "roomName": "Conference Room 1",
    "start": "2017-06-13T11:41:20.000Z",
  },
]
`;

exports[`selectors with bookings should be the same with bookingsByDate 1`] = `
Object {
  "2017-06-13T11:41:20.000Z": Array [
    Object {
      "end": "2017-06-13T11:41:20.000Z",
      "eventName": "Test Booking 01",
      "id": 0,
      "roomName": "Demo Room 01",
      "start": "2017-06-13T11:41:20.000Z",
    },
    Object {
      "end": "2017-06-13T11:41:20.000Z",
      "eventName": "Test Booking 02",
      "id": 1,
      "roomName": "Demo Room 02",
      "start": "2017-06-13T11:41:20.000Z",
    },
    Object {
      "end": "2017-06-13T11:41:20.000Z",
      "eventName": "Test Booking 03",
      "id": 2,
      "roomName": "Demo Room 03",
      "start": "2017-06-13T11:41:20.000Z",
    },
    Object {
      "end": "2017-06-13T11:41:20.000Z",
      "eventName": "Test Booking 04",
      "id": 3,
      "roomName": "Demo Room 04",
      "start": "2017-06-13T11:41:20.000Z",
    },
    Object {
      "end": "2017-06-13T11:41:20.000Z",
      "eventName": "Test Booking 05",
      "id": 4,
      "roomName": "Demo Room 05",
      "start": "2017-06-13T11:41:20.000Z",
    },
    Object {
      "end": "2017-06-13T11:41:20.000Z",
      "eventName": "Thanksgiving",
      "id": 5,
      "roomName": "Conference Room 1",
      "start": "2017-06-13T11:41:20.000Z",
    },
  ],
}
`;

exports[`selectors with bookings should be the same with currentSearchTerm 1`] = `""`;

exports[`selectors with bookings should be the same with loading 1`] = `true`;

exports[`selectors with bookings should be the same with selectBookings 1`] = `
Immutable.List [
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 01",
    "id": 0,
    "roomName": "Demo Room 01",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 02",
    "id": 1,
    "roomName": "Demo Room 02",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 03",
    "id": 2,
    "roomName": "Demo Room 03",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 04",
    "id": 3,
    "roomName": "Demo Room 04",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Test Booking 05",
    "id": 4,
    "roomName": "Demo Room 05",
    "start": "2017-06-13T11:41:20.000Z",
  },
  Object {
    "end": "2017-06-13T11:41:20.000Z",
    "eventName": "Thanksgiving",
    "id": 5,
    "roomName": "Conference Room 1",
    "start": "2017-06-13T11:41:20.000Z",
  },
]
`;

exports[`selectors with bookings should be the same with useFuzzySearch 1`] = `true`;

exports[`selectors with default states should be the same with allBookings 1`] = `
Immutable.List [
]
`;

exports[`selectors with default states should be the same with bookingsByDate 1`] = `Object {}`;

exports[`selectors with default states should be the same with currentSearchTerm 1`] = `""`;

exports[`selectors with default states should be the same with loading 1`] = `true`;

exports[`selectors with default states should be the same with selectBookings 1`] = `
Immutable.List [
]
`;

exports[`selectors with default states should be the same with useFuzzySearch 1`] = `true`;
